### PR TITLE
Add dm_control suite Hopper

### PR DESCRIPTION
**Pull request by Benjamin-eecs**
- [x] add dmc xml into third_party
- [x] load xml and assets
- [x] design class interface
- [x] domain randomization
- [x] reward calculation
- [x] align with dmc

### CODE REVIEW

**Comment by Trinkle23897** [comment]
```python
In [1]: from dm_control import suite

In [2]: e0 = suite.load("hopper", "hop", task_kwargs={"random": 0})
suite/hopper hop
mujoco/engine Physics __init__
mujoco/engine Physics after_reset
mujoco/engine Physics forward
suite/hopper task __init__
suite/base Task __init__
rl/control Environment __init__

In [3]: e0.reset()
rl/control Environment reset
rl/control Physics reset_context
mujoco/engine Physics reset
mujoco/engine Physics forward
suite/hopper task initialize_episode
suite/base Task initialize_episode
suite/base Task after_step
mujoco/engine Physics after_reset
mujoco/engine Physics forward
suite/hopper task get_observation
suite/hopper physics touch
Out[3]: 
TimeStep(step_type=<StepType.FIRST: 0>, reward=None, discount=None, observation=OrderedDict([('position', array([ 0.        ,  0.30670429,  0.22534578, -1.07342273,  1.46621727,
       -0.11992276])), ('velocity', array([0., 0., 0., 0., 0., 0., 0.])), ('touch', array([0., 0.]))]))

In [4]: a=e0.action_spec()

In [5]: e0.step(a.generate_value())
rl/control Environment step
suite/base Task before_step
mujoco/engine Physics set_control
mujoco/engine Physics step
suite/base Task after_step
suite/hopper task get_reward
suite/hopper physics height
suite/hopper physics speed
suite/hopper task get_observation
suite/hopper physics touch
rl/control Task get_termination
Out[5]: 
TimeStep(step_type=<StepType.MID: 1>, reward=0.0005784147110599047, discount=1.0, observation=OrderedDict([('position', array([-0.00249376,  0.3434393 ,  0.20517116, -1.0998353 ,  1.44573661,
       -0.12935117])), ('velocity', array([-0.04672792, -0.20055629,  2.9378971 , -1.61202892, -2.10634588,
       -1.64450644, -0.75285104])), ('touch', array([0., 0.]))]))
```

File: envpool/mujoco/dmc/mujoco_env.h [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    } else {
+      for (int i = 0; i < nstep; ++i) {
+        mj_step(model_.get(), data_.get());
+      }
+    }
+    mj_step1(model_.get(), data_.get());
+  }
+};
+
+}  // namespace mujoco
```
Comment (Trinkle23897) [inline envpool/mujoco/dmc/mujoco_env.h:194]: Just add a base class according to the discussion about #103
cc @Benjamin-eecs @kevinzakka

Comment (Trinkle23897) [reply envpool/mujoco/dmc/mujoco_env.h:194]: Not sure when TaskBeforeSubStep and TaskAftersSubStep are called in suite env.

**Review by Mehooz — COMMENTED** [review]
Overall I believe it is a good refactor. I leave the following comments:

1. _Dict-style indexing vs. hardcode._ In [dm_control hopper](https://github.com/deepmind/dm_control/blob/main/dm_control/suite/hopper.py), every attribute function will explicitly get the value with keys (like `sensor['torso_subtreelinvel']`), but in the current version, we hardcode the index of each attribute. An example is [Speed()](https://github.com/Benjamin-eecs/envpool/blob/dev-MujocoEnv/envpool/mujoco/dmc/hopper.h#L131). It is suitable for a task suite, but I am not sure if it will be ideal as an example for refactoring dm_control.
2. _Render._ I think an essential feature of dm_control task suite is visual input (see lots of papers are using dm_control with visual input settings). I am not sure about your roadmap, but it will be significantly improved if there is a[ rendering engine](https://github.com/deepmind/dm_control/tree/4e1a35595124742015ae0c7a829e099a5aa100f5/dm_control/_render) for all tasks. It might be a little bit hard to explore.

 But in general, I believe this module is promising for robot learning usage!

**Comment by Trinkle23897** [comment]
> Render.

This can be done with existing dm_control. My point is if you want to render for debugging, then use python dm_control; if you want faster training, then use envpool.

> Dict-style indexing vs. hardcode.

This is indeed an issue. I raise a similar issue https://github.com/deepmind/mujoco/issues/251
If we find a more elegant solution in the future, we'll update the codebase accordingly.

**Comment by Mehooz** [comment]
> > Render.
> 
> This can be done with existing dm_control. My point is if you want to render for debugging, then use python dm_control; if you want faster training, then use envpool.

Hi, here **render** means users may want to train a visual policy. Input is an image, and output is the motor command like the following papers did:
https://arxiv.org/abs/2107.09645
https://arxiv.org/abs/1912.01603
https://arxiv.org/abs/2107.00644

**Review by Trinkle23897 — APPROVED** [review]

**Pull request merged**